File: Editorial/1602_Find_Nearest_Right_Node_in_Binary_Tree/Python3/2_BFS_One_Queue_Sentinel.py
class Solution:
    def findNearestRightNode(self, root: TreeNode, u: TreeNode) -> TreeNode:
        if root is None:
            return None

        queue = deque([root, None,])
        while queue:
            curr = queue.popleft()

            # if it's the given node
            if curr == u:
                return queue.popleft()

            if curr:
                # add child nodes in the queue
                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
            else:
                # once the level is finished,
                # add a sentinel to mark end of level
                if queue:
                    queue.append(None)

File: Editorial/1601-1700/1602_Find_Nearest_Right_Node_in_Binary_Tree/CPP/2_BFS_One_Queue_Sentinel.cpp
class Solution {
public:
    TreeNode* findNearestRightNode(TreeNode* root, TreeNode* u) {
        if (root == nullptr) {
            return nullptr;
        }

        queue<TreeNode*> q;
        q.push(root);
        q.push(nullptr);

        while (!q.empty()) {
            TreeNode* curr = q.front();
            q.pop();

            if (curr != nullptr) {
                // if it's the given node
                if (curr == u) {
                    return q.front();
                }

                // add child nodes in the queue
                if (curr->left != nullptr) {
                    q.push(curr->left);
                }
                if (curr->right != nullptr) {
                    q.push(curr->right);
                }
            } else {
                // add a senitel to mark end of level
                if (!q.empty()) {
                    q.push(nullptr);
                }
            }
        }
        return nullptr;
    }
};
